llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -4560,8 +4560,15 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
     Tmp = ComputeNumSignBits(Op.getOperand(0), DemandedElts, Depth + 1);
     // SRA X, C -> adds C sign bits.
     if (const APInt *ShAmt =
-            getValidMinimumShiftAmountConstant(Op, DemandedElts))
+            getValidMinimumShiftAmountConstant(Op, DemandedElts)) {
       Tmp = std::min<uint64_t>(Tmp + ShAmt->getZExtValue(), VTBits);
+    } else {
+      KnownBits KnownAmt =
+          computeKnownBits(Op.getOperand(1), DemandedElts, Depth + 1);
+      if (KnownAmt.isConstant() && KnownAmt.getConstant().ult(VTBits))
+        Tmp = std::min<uint64_t>(Tmp + KnownAmt.getConstant().getZExtValue(),
+                                 VTBits);
+    }
```

Comment:
This seems like an unrelated change?

Reply:
If only... its to fix a thumb2 regression as it lowers v2i64 constant as bitcast(v4i32 constant)

Once this draft has addressed all the regressions I'll turn my attention to pulling out some of these changes.

Reply:
This is proving tricky to pull out - but I've confirmed that it doesn't cause any notable [compile time diff](https://llvm-compile-time-tracker.com/compare.php?from=7630379156ec08c9d7b1ea3c03c09e7dc89ef4ee&to=7a82854b3d28d02d37183ea91f09b1f7343c4bc4&stat=instructions:u) - as we fallback to ComputeKnownBits call which will call computeKnownBits on the shift amount anyway.

Reply:
Isn't there any helper simpler than computeKnownBits that can look through bitcasts to find a constant?

If you _are_ going to use computeKnownBits, why not use `KnownAmt.getMinValue()` instead of `KnownAmt.getConstant()`?

Reply:
Updated to getMaxValue() (for upper bound) + getMinValue() (for min sign extension) - the shift amount isn't just a `bitcast(v4i32 constant)` hidden constant, so we do need the abilities of computeKnownBits.

We could update getValidMinimumShiftAmountConstant (et. al) to return `std::optional<APInt>` to allow it to fallback to computeKnownBits, although that would mean the function would return a value that might not actual exist in the shift amount, I don't think we've used that property but it would still be a change.

Reply:
Actually I'm not even sure what the `ult` check is for, except perhaps to guard against `Tmp + getMinValue()` overflowing?

Reply:
Yes, its mainly just a sanity/overflow check (somebody always comes along with a i1024 fuzz test or something eventually that makes getZExtValue() assert or cause weird getLimitedValue() behaviour).

Using getMaxValue() was mainly to try and keep closer to the behaviour of getValidMinimumShiftAmountConstant which doesn't accept out of bounds shift amounts.